src/shared/install.c

```diff
@@ -2302,6 +2307,25 @@ int install_unit_unmask(
                 if (!unit_name_is_valid(*name, UNIT_NAME_ANY))
                         return -EINVAL;
 
+                if (!root_dir) {
+                        /* If root_dir is set, we don't care about kernel commandline or generators */
+                        InstallInfo info = {
+                                .name = *name,
+                                .install_mode = _INSTALL_MODE_INVALID,
+                        };
```

Comment:
hmppf, having two ways how UnitFileInstallInfo can be allocated/freed is messy. In particular destruction, since who knows what unit_file_search() might initialize in future.

Why not just allocate it on the heap and use the regular destructor? This is not a performance sensitive inner loop, so the copies/allocations this costs really shouldn't matter...

Reply:
Then in that case, it should probably be done the same way in the other place…